# Getting started with slurm


## Summary 

In this lab you will learn some of the basics of slurm and a few of the more advanced features that you might helpful in the future. We will be doing this lab on our class cluster but everything we are doing will also work on Stanford's Sherlock system (or any other slurm cluster) often with more interesting results because they are bigger more complex systems.


# Slurm basics


Slurm is a workload manager. Its jobs is to efficiently and fairly provide cluster resources. A slurm controlled cluster is ususally broken up into parittions where different sets of rules on use and are often composed of different hardware.  You use slurm resources by making a request to the slurm controller node for resources.  Your request is evaluated for priority based on what are referred to as "fairshare" rules. When there are free resources and you have the top priority you are granted access.

## sinfo

The sinfo command is very valuable way to find out about a cluster. Using the command without any options will give you a general idea about the cluster you are using.


In [1]:
!sinfo

PARTITION AVAIL  TIMELIMIT  NODES  STATE NODELIST
debug*       up   infinite      8  idle~ slurm-gp257-devel-compute-0-[4-11]
debug*       up   infinite      1    mix slurm-gp257-devel-compute-0-2
debug*       up   infinite      3  alloc slurm-gp257-devel-compute-0-[0-1,3]
cpu          up   infinite      5  idle~ slurm-gp257-devel-compute-1-[0-4]
t4           up   infinite     10  idle~ slurm-gp257-devel-compute-2-[0-9]
preempt      up   infinite     16 alloc# slurm-gp257-devel-compute-3-[2-17]
preempt      up   infinite     32  idle~ slurm-gp257-devel-compute-3-[18-49]
preempt      up   infinite      2  alloc slurm-gp257-devel-compute-3-[0-1]


In [2]:
!squeue
# I'm in the queue! :)

             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
              9624     debug     bash jdstitt_  R    2:26:05      1 slurm-gp257-devel-compute-0-0
             10627     debug     bash ellopes_  R      23:59      1 slurm-gp257-devel-compute-0-1
             10631     debug     bash ppettit_  R       9:47      1 slurm-gp257-devel-compute-0-2
             10636     debug     bash dneamati  R       4:23      1 slurm-gp257-devel-compute-0-3
           10637_1   preempt test_job tculliso CF       1:05      1 slurm-gp257-devel-compute-3-2
           10637_2   preempt test_job tculliso CF       1:05      1 slurm-gp257-devel-compute-3-3
           10637_3   preempt test_job tculliso CF       1:05      1 slurm-gp257-devel-compute-3-4
           10637_4   preempt test_job tculliso CF       1:05      1 slurm-gp257-devel-compute-3-5
           10637_5   preempt test_job tculliso CF       1:05      1 slurm-gp257-devel-compute-3-6
           10637_6   preempt test

By default the sinfo command gives you the names of the various partitions, the number of nodes in each partition, their status, and the name of the nodes in the various partitions.  This represents a very small amount of the sinfo commands capabilites. For example I can find the number of cores on a given node using the -n (for a specific node) and -O (for specific information) flags.

In [3]:
!sinfo -n slurm-gp257-devel-compute-0-0 -O CORES

CORES               
4                   


Below are some additional useful options.  Generic resources, Gres, refers to GPUs among other things.


| Option | Description|
| :--- | ------- |
| Cores | Number of cores per socket.|
| Disk  | Size of temporary disk space per node in megabytes.|
| FreeMem | The total memory, in MB, currently free on the node as reported by the OS. This value is for informational use only and is not used for scheduling.|
| Gres  | Generic resources (gres) associated with the nodes. |
| GresUsed | Generic resources (gres) currently in use on the nodes. |
| Memory | Size of memory per node in megabytes. |
| PreemptMode  | Preemption mode. |
|  Threads | Number of threads per core. |
| Time  |Maximum time for any job in the format "days-hours:minutes:seconds". |


In the cell below build a table describing the nodes on the cluster using the above commmand.

In [4]:
!sinfo -O Partition,Cores,Disk,FreeMem,Gres,GresUsed

PARTITION           CORES               TMP_DISK            FREE_MEM            GRES                GRES_USED           
debug*              4                   0                   23462-30758         (null)              gpu:0               
cpu                 30                  0                   237889-N/A          (null)              gpu:0               
t4                  8                   0                   52016-N/A           gpu:1               gpu:0               
preempt             30                  0                   237499-237898       (null)              gpu:0               


In [5]:
!sinfo -O Partition,Nodes,PreemptMode,Threads,Time

PARTITION           NODES               PREEMPT_MODE        THREADS             TIMELIMIT           
debug*              12                  OFF                 1                   infinite            
cpu                 5                   OFF                 1                   infinite            
t4                  10                  OFF                 1                   infinite            
preempt             50                  OFF                 1                   infinite            


The results of describing the nodes in the cluster are below. Note that the `FREE_MEM` entry changes on different runs of the code above.

| Partition | `debug` | `cpu` | `t4` | `preempt` |
| -- | -- | -- | -- | -- |
| Number of Nodes | 12 | 5 | 10 | 50 |
| Number of Cores per Node (`Cores`) | 4 | 30 | 8 | 30 |
| Temp Disk Size per Node (`Disk`) | 0 | 0 | 0 | 0 |
| Total Free Memory (`FreeMem`) in MB | 23462-30758 | 237889-N/A | 52016-N/A | 237499-237898 |
| Generic Resources (`Gres`) | (null) | (null) | gpu:1 | (null) |
| Generic Resources (`GresUsed`) | gpu:0 | gpu:0 | gpu:0 | gpu:0 |
| Preemption Mode (`PreemptMode`) | OFF | OFF | OFF | OFF |
| Number of Threads per Core (`Threads`) | 1 | 1 | 1 | 1 |
| Maximum Time for a Job (`Time`) | infinite | infinite | infinite | infinite |

# squeue

Thre squeue command gives us information about what is actually running on the cluster.  Often it useful to use the -p partiton option to reduce the infomation it produces on a cluster.


In [6]:
! squeue -p debug

             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
              9624     debug     bash jdstitt_  R    2:30:37      1 slurm-gp257-devel-compute-0-0
             10627     debug     bash ellopes_  R      28:31      1 slurm-gp257-devel-compute-0-1
             10631     debug     bash ppettit_  R      14:19      1 slurm-gp257-devel-compute-0-2
             10636     debug     bash dneamati  R       8:55      1 slurm-gp257-devel-compute-0-3
             10654     debug results1 ellopes_  R       0:01      1 slurm-gp257-devel-compute-0-2


In [7]:
# Compare to squeue alone
!squeue

             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
              9624     debug     bash jdstitt_  R    2:30:45      1 slurm-gp257-devel-compute-0-0
             10627     debug     bash ellopes_  R      28:39      1 slurm-gp257-devel-compute-0-1
             10631     debug     bash ppettit_  R      14:27      1 slurm-gp257-devel-compute-0-2
             10636     debug     bash dneamati  R       9:03      1 slurm-gp257-devel-compute-0-3
             10655     debug results2 ellopes_  R       0:02      1 slurm-gp257-devel-compute-0-2
              9623   preempt     bash jdstitt_  R    2:30:40      1 slurm-gp257-devel-compute-3-0
             10630   preempt     bash tculliso  R      14:06      1 slurm-gp257-devel-compute-3-1


In [18]:
!squeue --user=dneamati

squeue: error: Invalid user: dneamati

             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)


The results of the command gives you all the jobs that have been submitted to the cluster. It displays the jobid, the parition the job has been submitted, its status, how long it has been running, and the nodes the job is running on.  Below is the list of status possibilites.


|Status|Code|Explaination|
| ----- | ------ | ----- |
|COMPLETED|CD|The job has completed successfully.|
|COMPLETING|CG|The job is finishing but some processes are still active.|
|FAILED|F|The job terminated with a non-zero exit code and failed to execute.|
|PENDING|PD|The job is waiting for resource allocation. It will eventually run.|
|PREEMPTED|PR|The job was terminated because of preemption by another job.|
|RUNNING|R|The job currently is allocated to a node and is running.|
|SUSPENDED|S| A running job has been stopped with its cores released to other jobs.|
|STOPPED|ST|A running job has been stopped with its cores retained.|

The squeue and sinfo together provide a powerful tool in combination if you can't figure out why your job isn't starting (or fails)". Some complex examples include
- If you are requesting a job using a number of nodes you can see how many nodes are being currently used. Is it all by one user?
    - With `squeue`, we can see the users on each node (and the number of nodes). 
- If your job will not start, are you requesting more memory that in on the queue, more processors?
    - With `sinfo`, specifically the '-o' tag can be used list the memory and processor specifications for a node on a partition.
- Are other users using up all of the memory /GPUs currently?
    - With `sinfo`, we can see which nodes are allocated, mixed, or free. This can be especially useful within a partition. Specifically with the '-o' tag and `GRES` and `GRES_USED`, we can further check the GPU usage.


# salloc, sbatch, and srun

There are two general ways to request slurm resources, salloc and sbatch.  The easiest way to remember the differece is to use salloc when you want to do something interactivelty, sbatch when you are wanting a submit a job that runs eventually.  srun is used to launch jobs on these allocated resources but can be used directly (in whcih case it calls an salloc behind the scenes). 

The command you used to start this lab is a simple example

srun --pty --x11 /bin/bash

Here we requesteed a node on the default partition, with the default amount of memory, for the default amount of time, with a single core, said we wanted to run in terminal mode, and ran the command bash. 

In almost all cases you want to specify the resources you want to use it and the time you want to use them. The primary reason you want to specify the resources you need is the defults are often not approriate. You many need more memory, need more cores, etc. Almost as important is that it can often time give you an advatnage with the scheduler. If your job doesn't take much memory when another user is using most of the memory on a node, your job will start, while someone using the default might not. If the scheduler is trying to run a job that requires many cores, specifying your job will only run for a few minutes, the scheduler will run your job while waiting for other jobs to finish.





# Specifying job parameters (the basics)

We can specify slurm configuration parameters on the command line, or when using sbatch in the submission shell file by begining a line with #SBATCH before specify an option.

We can specify the parition using the -partition partName flag. The length of a job with --time=day-hours:minutes:seconds.  We can specify to use gpus by using the gres flag  --gres=gpu[[:type]:number].


## Specifying tasks

Slurm thinks of a job that has one or tasks. These task(s) run on one or more nodes.  We specify the number of tasks using --ntasks=ntasks.  If each parallel tasks uses multiple cores we can use --cpus-per-task=ncpus.  We can specify multiple tasks per core (--tasks-per-cpu=tasks) or node (--tasks-per-nodes). 

## Memory

There are a number of ways to specify. We can specify the total memory for a job with --mem=<size>units. We can also specify memory per cpu --mem-per-cpu=<size>[units] and per GPU --mem-per-gpu=<size>[units].



# Simple submission

In this part of the lab you will write a simple program to calculate pi using Leibniz’s formula.


X = 4 - 4/3 + 4/5 - 4/7 + 4/9 - ....

This series is never-ending, the more the terms this series contains, the closer the value of X converges to Pi value.

Use the next three cells to 

- write and save a python script that caclulates Pi using Leibniz formula. You should sum to 1000000 terms.
- write a sbatch that
    - submits a job to the compute partition
    - submits a time limit of 2 hours for the job
    - specifies one GB of memory
    - runs on a single core
- submits the jobs to the cluster


%%writefile leib.py
#!/usr/bin/env python3

# Do stuff

f=open("result1.txt","w")
f.write(res)
f.close()

%%writefile submit.sh
#SBATCH --ntasks=1


spack load python
python3 ./leib.py

!sbatch submit.sh

# Slurm advanced

Slurm also has a number of advanced options that can be quite useful.  

## Resubmitting

Clusters are always trying to maximize there usage.  As a result they often have a mechanism that allows unused resources used for free at or a reduced cost. In general you get to use these resources until someone who has/or is willing to pay/has paid a higher cost. At Stanford, the Sherlock clusters owner's queue operates in this manner. If you submit a job to the queue it will use all free resources if the owner of the resource requests those resources your job will be killed. GCP has the concept of preemitbible nodes, which opperate at a much lower cost, but the jobs can be killed at any time.

Many applications can take advantage of these potentially killed resources.  They are ideal for jobs that take a short time and don't use many cores. Slurm provides the option --requeue to enable the use of these resources.  Any job that doesn't complete is automatically resubmitted to the queue.

## Job arrays

There are many cases where you want to run a series of very similar jobs. Each job could be submitted in sequence but this can overload a slurm controller load.  Slurm job arrays provide an effective way to approach these jobs. The command line option --array offers a potential solution. You can submit an array in many forms:

- --array=0-31    # jobs 0 1 2 3 4 5 .. 31
- --array=1,3,5,7 # jobs 1 3 5 7
-  --array=1-7:2  # jobs 1 3 4 5 7

Each job will set the environmental variable SLURM_ARRAY_TASK_ID with the corresponding ID.


## Dependency

Another powerful feature of slurm is the ability to build dependencies. Basically guarantee that a task doesn't start until a given condition has been met. You can set a dependency using --dependency=<type:job_id[:job_id] where type is one of the following.


- after:jobid[:jobid...]	job can begin after the specified jobs have started
- afterany:jobid[:jobid...]	job can begin after the specified jobs have terminated
- afternotok:jobid[:jobid...]	job can begin after the specified jobs have failed
- afterok:jobid[:jobid...]	job can begin after the specified jobs have run to completion with an exit code of zero (see the user guide for caveats).



# Complex submission

A second way to calculate pi is to use random numbers. You can calculate pi by taking by 
 - choosing two random numbers between 0 and 1
 - checking whether to sum of the squares of those numbers is <= 1
 - 4 times the fraction of numbers that meet this criteria will be equal to pi


Your job is to

- write a program that follows the above procedure to estimate pi using a large number of tests
    - it should seed to random number generated based on reading the environmental variable SLURM_ARRAY_TASK_ID.
    - it should write the estimate to a file that uses the SLURM_ARRAY_TASK_ID in its name
- You should write a second program that reads a series of files with the name pattern above and averaes them and writes the result to result2.txt.
- You should write two slurm submission scripts
    - The First should submit to the preempt partition a job array of 1000 and resubmit if the job fails (make sure you test your code by running it on this node before submitting)
    - The second job should depend on the first job finishing with an exit code 0 befofre running the second python script. This job should run on debug partition

# Finishing up

Add this lab to your class github site after adding all of the files that you have created for this lab.